pkg/controller/gitlab/reconciler.go

```diff
@@ -482,9 +491,36 @@ func (a *applicationReconciler) reconcile(ctx context.Context, rr []resourceReco
 		return reconcileFailure, a.fail(ctx, reasonSyncingExternalDNS, err.Error())
 	}
 
-	// TODO(negz): Derive readiness from application state.
-	a.SetReady()
-	return reconcileSuccess, a.update(ctx)
+	if a.isReady(ctx) {
+		a.SetReady()
+		return reconcileSuccess, a.update(ctx)
+	}
+	return reconcileWait, a.update(ctx)
+}
+
+// TODO(negz): Consider more factors than whether the unicorn deployment is
+// available, i.e. whether the ingress controller and external DNS deployments
+// are ready.
+func (a *applicationReconciler) isReady(ctx context.Context) bool {
+	unicorn := a.GetName() + unicornResourceSuffix
+	r := &xpworkloadv1alpha1.KubernetesApplicationResource{}
+	n := types.NamespacedName{Namespace: a.GetNamespace(), Name: unicorn}
+	if err := a.client.Get(ctx, n, r); err != nil {
+		log.V(logging.Debug).Info("KubernetesApplicationResource not yet ready", "name", unicorn, "error", err)
+		return false
+	}
+
+	if r.Status.Remote == nil {
+		return false
```

Comment:
the other `return false` branches of this function debug log why it's not ready yet, do you want to debug log this `Remote == nil` condition too?

Reply:
I think that's a good idea, but this entire function is a big temporary hack so I'm just going to leave it as is for now.